### ISSUE TITLE
Manual cycle of SCC-DFTB with custom density matrix - projection between minimal and 3-21G* basis sets

### ISSUE THREAD

**Issue by AdamCoxson** [tbmalt/tbmalt#61]
Hi,

I want to take some density matrices at varying levels of convergence and pass them into an ML model I have built based on the 3-21G* basis to predict the Hamiltonian. I also want to do the same with DFTB to then compare the accuracy of both methods.

1) In my DFT atomic orbital based matrices I use the 3-21G* split-valence basis set (1s 2s, 3s, 2px, 2py, 2pz,  3px, 3py, 3pz, so 9 orbitals for Carbon, Nitrogen and Oxygen). I want to compare my results to DFTB, so I will use some minimal basis set with DFTB and later project it onto 3-21G* set. I have H, C,N, and O atoms in my molecules, what is the largest shell-dict/ basis set I can create? I'd like to get the 1s, 2s, 1p, 2p orbitals for CNO if thats possible? The shell dict might look something like this `{1: [0], 6: [0, 0, 1, 1], 7: [0, 0, 1, 1], 8: [0, 0, 1, 1]}`, but this doesn't work in the hamiltonian test script.

2) In tbmalt-development/tbmalt/physics/dftb/init I found the code to perform one cycle of SCC-DFTB. I would like to know how to use this with a custom guess of the density matrix (originally the guess would be in 3-21G* but I assume I have to project down to minimal? ... If thats possible). I figure I can take my density matrix, calculate the gross orbital populations and use these as q_in? Maybe a way to avoid 'down'-projection is to give the mulliken-charges per atom, so atom-resolved? but this loses even more information from the original density matrix.

Any thoughts on these two points would be appreciated.
            

`def _scc_cycle(self, q_in: Tensor) -> Tensor:
        """Perform a single self-consistent charge cycle.

        This method performs a single Self-Consistent Charge cycle (SCC). Using
        ``q_in`` as the initial guess during the first cycle and as the mixed
        charges in all subsequent cycles.

        Arguments:
            q_in: Input charges are provided via the ``q_in`` argument. This
                can be viewed as the initial guess for the first cycle and as
                the mixed charges from the previous step in all subsequent
                cycles.

        Returns:
            q_out: New updated charges, as computed following the SCC step.

        Notes:
            It is important to note that this method is intended to modify and
            update the class and its attributes during the SCC cycle. The newly
            computed charges are only returned to facilitate ease of use.

            The charges ``q_in`` and ``q_out`` may be either shell or atom
            resolved, but must match up with that as defined by the orbs
            attribute `shell_resolved`.
        """

        # Construct the shift matrix
        shifts = torch.einsum(
            '...i,...ij->...j', q_in - self.q_zero_res, self.gamma)
        shifts = prepeat_interleave(shifts, self.orbs.orbs_per_res)
        shifts = (shifts[..., None] + shifts[..., None, :])

        # Compute the second order Hamiltonian matrix
        self._hamiltonian = self.core_hamiltonian + .5 * self.overlap * shifts

        # Obtain the eigen-values/vectors via an eigen decomposition
        self.eig_values, self.eig_vectors = eighb(
            self.hamiltonian, self.overlap, **self._solver_settings)

        # Scaled occupancy values
        s_occs = torch.einsum(
            '...i,...ji->...ji', torch.sqrt(self.occupancy), self.eig_vectors)

        # Density matrix
        self.rho = s_occs @ s_occs.transpose(-1, -2).conj()

        # Compute and return the new
        return _mulliken(self.rho, self.overlap, self.orbs)
`

**Comment by mcsloy** [comment]
Hello Adam,

I have tried to address the three main points that you raised as best I can. Let me know if you require more information or examples, and I will do what I can to help. I may not have addressed all of your points in this comment, but my response was already becoming a little dense. So we can expand the discussion as necessary.

### Q1) "*I want to take some density matrices at varying levels of convergence...*"

For the *non-batch* case, one can simply hijack the `_scc_cycle` call to copy over the density matrix into an appropriate record tensor like so:
```py
class Dftb2WithRhoHistory(Dftb2):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.rho_history: Optional[Tensor] = None

    def _scc_cycle(self, q_in: Tensor) -> Tensor:
        # Invoke the parent's `_scc_cycle` method to get the new charges &
        # trigger an update of the `rho` attribute.
        charges = super()._scc_cycle(q_in)

        # Get a detached copy of the newly computed rho matrix.
        rho = self.rho.detach().clone().unsqueeze(0)

        # Append this to the `rho_history` attribute.
        self.rho_history = (rho if self.rho_history is None
                            else torch.cat((self.rho_history, rho)))

        # Finally return the charges
        return charges

    def reset(self):
        super().reset()
        self.rho_history = None
```

In the above example, the attribute `rho_history` will store the density matrix as computed at each step of the SCC cycle. However, it is important to note that this will **not** work on batches of systems; only single system calculations. This is because batches are systematically culled during the SCC cycle as the molecules within them converge. This means that one might end up with 10 density matrices for the 1st system, 5 for the 2nd, 8 for the 3rd, and so on. Furthermore, as the only function that knows which systems converge and when is the `forward` method, the implementation will not be clean. However, if density matrix logging for batched systems is something that you require, I can create a custom solution for you.

### Q2) "*....what is the largest shell-dict/ basis set I can create?*"

In short there is no explicit limit placed upon the size of the shell dictionary or its contents. However, this must match up **exactly** with what the feeds/parameter-sets are able to provide (explanation given below). 

Generally speaking, the `OrbitalInfo` class is *descriptive* rather than *prescriptive* in that it is not intended control anything itself. Its primary purpose is to give different parts of the code a common way to access information like i) what is the anticipated size of the H/S matrix, ii) which blocks/sub-blocks/elements of such matrices are associated with a particular atom, iii) what shells are expected to be present on the n'th atom in a molecule, and so on. Thus `shell_dict` is not normally a free parameter.

For example, if one were to use the [auorg-1-1](https://dftb.org/parameters/download/auorg/auorg-1-1-cc) parameter set to perform an SCC-DFTB calculation, the shell dictionary must be {1: [0],  ...} (I have omitted non-hydrogen species here for the sake of brevity). One cannot freely change the shell specification for hydrogen to say {1: [0, 1], ...}. Not because the `OrbitalInfo` class forbids it, but because the parameter set simply does not contain such information. Setting the shell dictionary to `{1: [0, 1], ...}` informs TBMaLT that hydrogen atoms have a "p" orbital. TBMaLT would then expect the feed objects representing the auorg-1-1 parameter set to be able to provide information about the hydrogen's "p" orbital. Which is impossible, because the parameter set contains no such information, resulting in an error.

I apologies if I have misunderstood your question. Please let me know if this is the case and I will update this response accordingly.

### Q3) "*.... I found the code to perform one cycle of SCC-DFTB \[\_scc\_cycle\]. I would like to know how to use this with a custom guess of the density matrix ... I figure I can take my density matrix, calculate the gross orbital populations and use these as q_in?...*"

Within DFTB the charges are traditionally used as the starting point rather than the density matrix. So the best course of action would be to calculate the populations as you suggest. Custom initial starting guesses for the charges can be provided using the cache feature like so:

```py
dftb2_calc(geometry, orbital_info, cache={"q_initial": starting_guess})
```

Admittedly, the above example assumes you want to perform a full SCC-DFTB calculation. If you want to perform only a single SCC step, then you could, at least in theory, manually invoke the `_scc_cycle` method. However, I would advise *extreme caution* if you choose to go this route. This method is flagged for internal use only, and so no guarantees are made that this will behave as intended. You would need to ensure that all of the internal requirements of this function are met; i.e. necessary attributes have been assigned, and so on. That being said, I just ran a quick test and it *seems* to work.

A safer, if somewhat obscure looking, approach might look a little something like this:
```py
# Set max number of scc-steps to 0 so that only a single "clean-up" SCC  
# step is performed.
dftb_calculator.max_scc_iter = 0  
# This will always trigger a convergence error; thus the error suppression  
# flag must be set.
dftb_calculator.suppress_SCF_error = True  
# Run the DFTB calculator for a single step using a custom starting guess  
# for the charges.  
dftb_calculator(geometry, orbital_info, cache={"q_initial": starting_guess})
```
Here, one is effectively running an SCC-DFTB calculation for only a single step using a custom value for the initial starting charges.

**Comment by AdamCoxson** [comment]
Okay, so heres some rough pseudo-steps I'll try over the next week or so

Some ingredients:
I am only interested in the non-batch case
I think I can do matrix basis projection using [pySCF code](https://pyscf.org/_modules/pyscf/scf/addons.html#project_dm_nr2nr)
I will try to use the Ob2-1-1 DFTB parameter set which is H s orbitals and CNO p orbitals which I think corresponds to {1: [0], 6: [0, 1], 7: [0, 1], 8: [0, 1]}

- Take my density matrix guess which is in the 3-21G* basis. Project it onto the Ob2-1-1 basis and then take the orbital-resolved mulliken populations (mPop).
- Pass mPop into dftb_calculator(q_init=mPop, max_iter=0)
- Extract the new density matrix and project back into 3-21G* basis.
- Compare DFTB prediction to my ML prediction.

I reckon the projection won't be so simple but will have to try it first. This code should also allow me to do a full DFTB run by changing max_iter.
